api/workAllocation/caseWorkerUserDataCacheService.ts

```diff
@@ -13,81 +13,136 @@ import { CachedCaseworker, LocationApi } from './interfaces/common';
-      const userResponse = await handleUsersGet(getUsersPath, req);
-      // TODO: Response will be cached eventually via API so caching below should be removed eventually
-      cachedUsers = getUniqueUsersFromResponse(userResponse);
+
+      let headers;
+      if (!req) {
+        await getAuthTokens();
+        headers = getRequestHeaders();
+      }
+
+      // Fetch all user data with pagination
+      const allUsers = await fetchAllUserPages(jurisdictions, req, headers);
+
+      // Process and cache the results
+      state.cachedUsers = getUniqueUsersFromResponse(allUsers);
     } else {
-      refreshRoles = false;
+      state.refreshRoles = false;
     }
-    return cachedUsers;
+
+    return state.cachedUsers;
   } catch (error) {
-    if (cachedUsers) {
-      return cachedUsers;
+    // Handle errors appropriately
+    if (state.cachedUsers && state.cachedUsers.length > 0) {
+      // Return cached data if available despite error
+      return state.cachedUsers;
+    }
+
+    // Reset cache on critical errors with no request handler
+    if (!next) {
+      state.timestamp = null;
+      state.refreshRoles = false;
+      return null;
     }
+
     next(error);
   }
 }
 
-// Note: May not be needed once API caching is in effect
-export async function fetchNewUserData(): Promise<StaffUserDetails[]> {
-  try {
-    // first ensure that there is no immediate re-caching
-    timestamp = new Date();
-    refreshRoles = true;
-    await getAuthTokens();
-    // add the tokens to the request headers
-    const caseworkerHeaders = getRequestHeaders();
-    const jurisdictions = getConfigValue(STAFF_SUPPORTED_JURISDICTIONS);
-    cachedUsers = [];
-    const getUsersPath: string = prepareGetUsersUrl(baseCaseWorkerRefUrl, jurisdictions);
-    const userResponse = await handleNewUsersGet(getUsersPath, caseworkerHeaders);
-    cachedUsers = getUniqueUsersFromResponse(userResponse);
-    return cachedUsers;
-  } catch (error) {
-    if (cachedUsers) {
-      return cachedUsers;
+// Helper function to fetch all pages of user data
+async function fetchAllUserPages(
+  jurisdictions: string,
+  req?: EnhancedRequest,
+  headers?: Record<string, string>
+): Promise<StaffUserDetails[]> {
+  let pageNumber = 0;
+  let hasMoreData = true;
+  let allUsers: StaffUserDetails[] = [];
+  const pageSize = parseInt(getConfigValue(CASEWORKER_PAGE_SIZE));
+
+  while (hasMoreData) {
+    const getUsersPath = prepareGetUsersUrl(baseCaseWorkerRefUrl, jurisdictions, pageNumber);
+
+    // Use appropriate method based on context
+    const userResponse = req
+      ? await handleUsersGet(getUsersPath, req)
+      : await handleNewUsersGet(getUsersPath, headers);
+
+    if (!userResponse || userResponse.length === 0) {
+      hasMoreData = false;
+    } else {
+      allUsers = [...allUsers, ...userResponse];
+
+      // Check if this is the last page
+      if (userResponse.length < pageSize) {
+        hasMoreData = false;
```

Comment:
I think this will go slightly wrong if the last page of results has exactly pageSize results, ie the total number of results is an exact multiple of page size. Probably it will just do a one more request than is needed, for page n+1 of n, but check if this causes an error which could make the whole thing appear to have failed